### ISSUE TITLE
Version 1.0

### ISSUE THREAD

**Issue by GitRon** [EmilStenstrom/django-components#433]
Hi @EmilStenstrom 

the package looks amazingly well maintained and has a full feature set. Would you think that maybe it's time for the first major release?

Best from Cologne  
Ronny

# Acceptence criteria
* Version bump to 1.0.0
* PyPI metadata "5 Production/Stable" is set

**Comment by GitRon** [comment]
Supplemental: A 0.x release might disencourage some devs to use this package.

**Comment by EmilStenstrom** [comment]
I'd like to get some things in order before we release 1.0.0. Here's the current set of issues that I think are blocking that release: https://github.com/EmilStenstrom/django-components/milestone/1

**Comment by GitRon** [comment]
Oh, didnt know! Seems like a good plan! 💪

**Comment by JuroOravec** [comment]
It's been a long journey in the past few months. But there's only a few more features, and I think we can come back to the topic of v1 🎉 . Features that I want to get in before then is:
- [Fix template caching](#589)
- [Remove safer_staticfiles](#647)
- [Introduce JS / CSS vars, client-side dependency manager, scoped CSS (they are all connected)](#622)
- [HTML fragments support](#635)
- Plugin system

Might seem like a lot, but the code is written up for the first 3 items already, so it's not that far.

**Comment by EmilStenstrom** [comment]
@JuroOravec When you have the time, please update this list with what you think should be in 1.0!

**Comment by JuroOravec** [comment]
@EmilStenstrom Sure, I'll write it as general areas / groups of related features:

- [ ] 1. [feat] JS / CSS variables, ~scoped CSS~

   - See:
     - https://github.com/django-components/django-components/issues/622
     - https://github.com/django-components/django-components/issues/836
   - Status:
     - [ ] CSS variables - missing just tests and documentation
     - [ ] JS variables - implemented locally, needs tests and documentation
     - [ ] ~Scoped CSS - implemented 50%. What's remaining is to apply CSS scoping also to the slots that were defined within `{% fill %}` tags to make it work like in Vue.~
     - [ ] Allow to export inlined JS / CSS
         - Note: The proper solution here would be to allow to export the inlined JS / CSS, to manage it as separate files:
            - Command to export the inlined JS / CSS (or part of `collectstatic`)
            - Generate a "manifest" file - JSON that maps which JS / CSS belong to which component
            - Then somehow make the components NOT render their own JS / CSS when the manifest is detected.
            - !!Note!! - Not sure how feasible this will be tho, if we want to also have `Component.js_render()`.
         - Closes https://github.com/django-components/django-components/issues/689
     - [ ] Allow to define how to wrap Component's JS / CSS in `<script>` / `<style>`  with `Component.js_render()` and `Component.css_render()`
         - Part of https://github.com/django-components/django-components/issues/932
      - [ ] Refactor `Component.Media` into extension so users can globally define how to render `<script>` and `<link>` tags for JS / CSS from `Component.Media` ([#1144](https://github.com/django-components/django-components/issues/1144))
         - Part of https://github.com/django-components/django-components/issues/932
      - [ ] Shorten the placeholder tags to `{% components_js %}` and `{% components_css %}`?

- [ ] 2. [feat] Finish and polish the extra template tag syntax

   - See:
     - https://github.com/django-components/django-components/issues/935,
     - https://github.com/django-components/django-components/issues/872,
     - https://github.com/django-components/django-components/discussions/900
   - Status:
     - Right now one can write literal lists and dicts inside the `{% component %}` tags, e.g. `{% button items=[1, 2, 3] %}`. But right now it's implemented in Python and slow (adds about 1s on first render in my work project).

     - I wrote a Rust parser for this (Note: this is DIFFERENT parser than [djc-core-html-parser](https://github.com/django-components/djc-core-html-parser)). Still local, haven't pushed it yet.

        - NOTE: This is basically an AST parser, and superset of Django's syntax. So this could be used in the future to add IDE support for extra template features introduced by django-components.

     - So what remains is to update djc to use the Rust AST parser, add tests, and document all the extra features (not all are documented at the moment).

- [ ] 3. [cleanup] Component's API surround templates / rendering

    - See https://github.com/django-components/django-components/issues/1085
    - Status:
       - As mentioned in https://github.com/django-components/django-components/issues/1085:
          > ### Prep for future features
          > I've mentioned it on multiple topics, but having the ability to dynamically get a template for a component (with `Component.get_template_name()` and `Component.get_template()`) causes issues for:
          > 
          > * Django-cotton syntax, and overall the ability to pre-process Templates via extension hooks ([Comparision with django-cotton #552](https://github.com/django-components/django-components/discussions/552), [Django-cotton-like html tag syntax for django-components #794](https://github.com/django-components/django-components/issues/794))
          > * Ability to optimize templates ([[v2] Mark constant context vars for 50% perf gain #1083](https://github.com/django-components/django-components/issues/1083))
          > * Ability to write the templates in languages other than Django / HTML, and "pre-compile" them into Django templates.
        - [X] Deprecating `get_template_name()` and `get_template()` ([#1222](https://github.com/django-components/django-components/pull/1222))
        - [X] Allow components with no templates ([#1222](https://github.com/django-components/django-components/pull/1222))
        - [ ] `ErrorFallback` component (mentioned in the same issue)
        - [x] `Compont.on_render()` for intercepting what gets rendered. This would also make the implementation of the "dynamic" component cleaner.
        - [x] Deprecate manual Component instantiation and instead instantiate inside Component.render() [#1186](https://github.com/django-components/django-components/pull/1186)

- [x] 4. [cleanup] Loading of Templates.

   - Note: While 3. focused on the API, this point focuses on the internal handling of Templates.
   - See https://github.com/django-components/django-components/issues/901
   - Status:
       - Implemented.
       - [X] Currently I find it hard to follow / reason about the logic that loads templates (`Component._get_template`). IMO it deserves to be in its own file. ([#1222](https://github.com/django-components/django-components/pull/1222))


- [x] 5. [cleanup] Resolve rendering of JS / CSS dependencies when Template is rendered with Django's API.

   - See https://github.com/django-components/django-components/issues/968#issuecomment-2653965260, and https://github.com/django-components/django-components/pull/959#issuecomment-2824435761.
   - Status:
     - Implemented.
     - NOTE: In an ideal scenario, we'd come up with a way of rendering components to be able to insert one into another without needed to set `render_dependencies=False`. But it feels like that's outside of the scope of v1.
     - I'm also thinking of renaming the kwarg from `render_dependencies=False` to `finalize=False`. But that's just eye candy.

- [x] 6. [cleanup] Usage around slots

   - Status:
     - [x] Change signature of slot functions - https://github.com/django-components/django-components/issues/1096
     - [x] Document accessing slots in code - https://github.com/django-components/django-components/issues/1095
     - [x] Expose `Slot.contents` for caching - https://github.com/django-components/django-components/issues/1164#issuecomment-2854682354
     - [x] Allow passing slots to fill tags as `{% fill "footer" slot=my_slot / %}` - https://github.com/django-components/django-components/issues/1171
     - [x] Move slot escaping from inside `Component.render()` to `Slot` instances
     - [x] Replace `component_vars.is_filled` with `component_vars.slots` - https://github.com/django-components/django-components/issues/1174
     - [x] Implement extension hook that's called when slot is rendered. (This would allow to move the debug highlighting into an extension).
     - [x] Allow slots to pass arbitrary data (e.g. via `Slot.meta` dictionary) to enable CSS scoping.
     - [x] Replace `Slot.source` replace with `Slot.fill_node`
   - Also see
     - https://github.com/django-components/django-components/discussions/1091

- [ ] 7. [cleanup] Review the flow for sharing component libraries.

   - See https://django-components.github.io/django-components/0.139.1/concepts/advanced/authoring_component_libraries/
   - Status:
     - Right now a single component can be registered under different names and in different registries.
     - However, I'm wondering whether it wouldn't make sense to simplify this, and set the component name once on the Component class itself. Then there would be less ambiguity, because the same component would always have the same name in the templates.
     - Also, when writing a test standalone component (https://github.com/JuroOravec/djc-heroicons), I found it more intuitive to "share" components by simply registering individual components to the Registry I provided.
     - This is different from what is recommended in the docs, which says to put them all in a new Registry, and share them that way.
     - The problem is that when we create multiple Registries, we might accidentally register multiple components under the same name, and thus same template tag.
     - While setting component name directly on the component might be better left for [v2](https://github.com/django-components/django-components/issues/1141), I think for v1 we should implement some logic that checks the component names across all registries, so that we raise an error if multiple registries use the same template tags.
     - [ ] Make `@register` optional with default component template name set by `Component.name` (https://github.com/django-components/django-components/issues/1195)

- [ ] 8. [cleanup] Misc
   - [ ] All the [documentation issues](https://github.com/django-components/django-components/issues?q=sort%3Aupdated-desc%20is%3Aissue%20is%3Aopen%20label%3Adocumentation)
   - [ ] Simplify `{% html_attrs %}` - Shorten to `{% attrs %}`; make `default` arg keyword-only; accept many positional args (similar to Tetra)
   - [ ] Add `Component.parent`, `Component.root`, `Component.ancestors`

So that leaves the stuff that's outside of the scope / nice to have:
- Support for TypeScript, Sass, etc
- https://github.com/django-components/django-components/issues/897
- https://github.com/django-components/django-components/issues/1118
- [django-debug-toolbar integration](https://github.com/django-components/django-components/issues/957)
- [Exporting JS / CSS files to static files](https://github.com/django-components/django-components/issues/689)
- [Django system checks](https://github.com/django-components/django-components/issues/882)
- https://github.com/django-components/django-components/issues/835
- [HTML syntax for components](https://github.com/django-components/django-components/issues/794) (AKA django-cotton style)
- [AlpineJS integration](https://github.com/django-components/django-components/issues/644)
- Vue integration.
- [Scoped CSS extension](https://github.com/django-components/django-components/issues/1230)

**Comment by JuroOravec** [comment]
Btw @EmilStenstrom I wrote out more details for the things to be done around slots, this is what I want to focus on next.

**Comment by JuroOravec** [comment]
(Update copied from https://github.com/django-components/django-components/pull/1221)

Btw quick update, so there's 7 PRs (including this one) that I want to get it before releasing v0.140:

1. `slot.source` and `slot.extra`
2. Deprecate `Component.get_template_name()` and `Component.get_template()`, deprecate template caching, associate template with Comp cls, allow `template = None`
    - Will close https://github.com/django-components/django-components/issues/901
    - Will close https://github.com/django-components/django-components/issues/709
3. Component subclassing should not inherit parent's template/js/css when child's value is set to `None`
4. `Component.on_render()` and component error handling
    - Will close https://github.com/django-components/django-components/issues/1085
5. Add hooks `on_template_loaded()` (str), `on_template_compiled()` (Template), `on_js_loaded()` (str), `on_css_loaded()` (str)
6. Add `slot_node` to `on_slot_rendered()` hook, so we can find out which component the slot node came from (again part towards scoped CSS extension)
7. Deprecate `Component.input`, add `Component.raw_args`, `Component.raw_kwargs`, `Component.raw_slots`
    - Will close https://github.com/django-components/django-components/issues/1206

After that and after releasing v0.140, there will be only 3 mental groups* remaining for v1:
- JS / CSS variables, and other stuff around `dependencies.py`
- Template tag parsing
- Component registry / registering components / etc

\* Plus some stray features:
- Refactoring `Component.Media` into extension, so media class methods could be overriden globally through `extensions_defaults`
- `ErrorFallback` component
- `Component.parent`, `Component.root` attributes, to be able to walk up the component tree

Btw after v0.140 I will also take a break for a month or so, I'll be moving to Prague in June / July 🎉 🍺

**Comment by joeyjurjens** [comment]
Do you guys have any idea regarding breaking changes once v1.0 is reached? It seems like (from above comment) that those things are left:

> After that and after releasing v0.140, there will be only 3 mental groups* remaining for v1:
> 
> JS / CSS variables, and other stuff around dependencies.py
> Template tag parsing
> Component registry / registering components / etc

With those, I can't think of much breaking changes regarding the general usage of components itself.

I am asking because I am considering using this library on a new system, but if big breaking changes are expected then maybe I can better wait until v1.0 is reached. However, if there aren't then I can start working on it now and then I don't have to rewrite regular django templates once django-components reaches v1.0 which would also be a lot of work 😄.

**Comment by EmilStenstrom** [comment]
@joeyjurjens We are trying to be considerate of breaking changes already, so I would say no need to wait for 1.0. I doubt there would be any major changes that would require you to rewrite all your templates... And if there are, a few well-written regex replaces usually does the thing.
